pkg/container-collection/tracerconnect/connect.go

```diff
@@ -0,0 +1,122 @@
+	Tracer        NetworkingTracer[Event]
+	Resolver      containercollection.ContainerResolver
+	Selector      containercollection.ContainerSelector
+	EventCallback func(*containercollection.Container, Event)
+	Base          func(eventtypes.Event) Event
+}
+
+// ConnectNetworkingTracerToContainerCollection connects a networking
+// tracer to the container collection package. It creates the needed
+// logic to call the Attach() function on the tracer each time a
+// container is created and to call Detach() each time the container is
+// removed. Callers must call args.Resolver.Unsubscribe(key) to remove
+// the connection.
+func ConnectNetworkingTracerToContainerCollection[Event any](
+	args *ConnectNetworkingTracerToContainerCollectionArgs[Event],
+) error {
+	return connectNetworkingTracerToContainerCollectionInternal(
+		args.Key, args.Tracer, args.Resolver, args.Selector, args.EventCallback,
+		args.Base,
+	)
+}
+
+func connectNetworkingTracerToContainerCollectionInternal[Event any](
+	key string,
+	tracer NetworkingTracer[Event],
+	resolver containercollection.ContainerResolver,
+	selector containercollection.ContainerSelector,
+	eventCallback func(*containercollection.Container, Event),
+	base func(eventtypes.Event) Event,
+) error {
+	netnsHost, err := containerutils.GetNetNs(os.Getpid())
+	if err != nil {
+		return fmt.Errorf("getting network namespace: %w", err)
+	}
+	genKey := func(container *containercollection.Container) string {
+		if container.Netns == netnsHost {
+			return "host"
+		}
+
+		key := container.Namespace + container.Podname
+		// for local-gadget as container.Namespace and
+		// container.Podname can be empty
+		if key == "" {
+			key = container.ID
+		}
+
+		return key
+	}
+
+	attachContainerFunc := func(container *containercollection.Container) {
+		cbWithContainer := func(ev Event) {
+			eventCallback(container, ev)
+		}
+		err := tracer.Attach(genKey(container), container.Pid, cbWithContainer)
+		if err != nil {
+			msg := fmt.Sprintf("failed to trace container: %q: %s", container.Name, err)
+			cbWithContainer(base(eventtypes.Err(msg)))
+			return
+		}
+		cbWithContainer(base(eventtypes.Debug("tracer attached")))
+	}
+
+	detachContainerFunc := func(container *containercollection.Container) {
+		err := tracer.Detach(genKey(container))
+		if err != nil {
+			msg := fmt.Sprintf("failed to stop tracing container: %q: %s", container.Name, err)
+			eventCallback(container, base(eventtypes.Err(msg)))
+			return
+		}
+		eventCallback(container, base(eventtypes.Debug("tracer detached")))
+	}
+
+	containers := resolver.Subscribe(
+		key,
+		selector,
+		func(event containercollection.PubSubEvent) {
+			switch event.Type {
+			case containercollection.EventTypeAddContainer:
+				attachContainerFunc(event.Container)
+			case containercollection.EventTypeRemoveContainer:
+				detachContainerFunc(event.Container)
+			}
+		},
+	)
+
+	for _, container := range containers {
+		attachContainerFunc(container)
+	}
```

Comment:
With this change, we are not calling `tracer.Detach` for these containers anymore (for the ones that stay alive for the full life of the tracer). Therefore, we need to be sure that `tracer.Close` is releasing all the links created in the attach for these already existing containers. Maybe a test to be added to the gadget unit tests:
1. Call `tracer.NewTracer`
2. Call `Subscribe` and store the existing containers
3. Call `tracer.Attach` for those existing containers
4. Call `tracer.Close`
5. Verify all the links were released.

Do you think it would make sense?

Reply:
I think it makes sense but I prefer to do this in another PR later on.

Reply:
Yes, of course 🙂